File: packages/backend/agents/org_agent.py
"""
Org Analysis Flow — Python wrapper (not a single agent).
Orchestrates: List-Hosts Agent → batch fan-out → Summarize Agent.
"""
import asyncio
import logging
from agents.list_hosts_agent import run_list_hosts_agent
from agents.host_batch_agent import run_host_batch_agent
from agents.summarize_agent import run_summarize_agent
from agents.tools.dynamodb_tools import read_host_list

logger = logging.getLogger(__name__)

BATCH_SIZE = 10


async def run_org_analysis(tenant_id: str, okta_token: str, run_id: str) -> None:
    """
    Full org analysis flow for one tenant:
    1. List-Hosts Agent: discover hosts, write to DynamoDB
    2. Python: read host list, fan out batches
    3. Host Batch Sub-Agents: analyze hosts in parallel (asyncio.gather)
    4. Summarize Agent: compute org summary, write to DynamoDB
    """
    logger.info(f"[org_analysis:{tenant_id}] Starting host discovery")

    # Invocation 1: List-Hosts Agent
    await run_list_hosts_agent(tenant_id, okta_token, run_id)

    # Python step: read host list from DynamoDB
    hosts = read_host_list(tenant_id, run_id)
    if not hosts:
        logger.warning(f"[org_analysis:{tenant_id}] No hosts found — skipping batch analysis")
        # Still run summarize to write an empty summary
        await run_summarize_agent(tenant_id, okta_token, run_id)
        return

    logger.info(f"[org_analysis:{tenant_id}] Found {len(hosts)} hosts, fanning out batches")

    # Python step: fan out batches of BATCH_SIZE
    batches = [hosts[i:i + BATCH_SIZE] for i in range(0, len(hosts), BATCH_SIZE)]
    total_batches = len(batches)

    await asyncio.gather(*[
        run_host_batch_agent(
            tenant_id=tenant_id,
            hosts=batch,
            okta_token=okta_token,
            run_id=run_id,
            batch_index=i,
            total_batches=total_batches,
        )
        for i, batch in enumerate(batches)
    ])

    logger.info(f"[org_analysis:{tenant_id}] All batches complete, running summarize")

    # Invocation 2: Summarize Agent
    await run_summarize_agent(tenant_id, okta_token, run_id)

    logger.info(f"[org_analysis:{tenant_id}] Org analysis complete")

File: packages/agent/src/agents/org-agent.ts
import { runListHostsAgent } from "./list-hosts-agent.js";
import { runHostBatchAgent } from "./host-batch-agent.js";
import { runSummarizeAgent } from "./summarize-agent.js";
import { runMetricPrefetch, runHostMetadataPrefetch } from "./metric-prefetch-agent.js";
import { readHostList } from "../tools/dynamodb.js";
import { isAborted } from "../tools/abort-registry.js";

const BATCH_SIZE = 15;       // 15 hosts per batch — process_batch_tool handles all 15 in parallel server-side
const BATCH_CONCURRENCY = 5; // 5 concurrent batches — each batch is 2 LLM turns with tiny context, no rate limiting

export async function runOrgAnalysis(
  tenantId: string,
  runId: string
): Promise<void> {
  console.log(`[org_analysis:${tenantId}] Starting host discovery`);

  await runListHostsAgent(tenantId, runId);

  if (isAborted(runId)) {
    console.log(`[org_analysis:${tenantId}] Run aborted after host discovery — stopping`);
    return;
  }

  const hosts = await readHostList(tenantId, runId);
  if (hosts.length === 0) {
    console.warn(`[org_analysis:${tenantId}] No hosts found — skipping batch analysis`);
    await runSummarizeAgent(tenantId, runId);
    return;
  }

  // ── Bulk pre-fetch: metrics + host metadata ────────────────────────────────
  // Both run in parallel before batch agents start.
  //
  // Metric pre-fetch: collapses N_hosts × 23 per-host Datadog calls into
  //   23 org-wide calls (≤1000 hosts) or N_chunks × 23 calls (>1000 hosts).
  //   Batch agents call get_prefetched_metrics_tool — always a cache hit.
  //
  // Host metadata pre-fetch: collapses N_hosts search_datadog_hosts MCP calls
  //   into ceil(N/1000) REST API calls (GET /api/v1/hosts, 1000/page).
  //   Batch agents call get_prefetched_host_metadata_tool — no MCP needed for Step A.
  //   Falls back to search_datadog_hosts only for individual cache misses (rare).
  //
  // If either pre-fetch fails it is non-fatal — batch agents degrade gracefully:
  //   metric failure  → all hosts get efficiency_label = "unknown"
  //   metadata failure → batch agents fall back to per-host search_datadog_hosts
  console.log(`[org_analysis:${tenantId}] Starting metric + host metadata pre-fetch for ${hosts.length} hosts`);
  const [metricResult, metadataResult] = await Promise.allSettled([
    runMetricPrefetch(tenantId, runId, hosts),
    runHostMetadataPrefetch(tenantId, runId),
  ]);

  if (metricResult.status === "fulfilled") {
    const { metricsStored, hostsWithData } = metricResult.value;
    console.log(`[org_analysis:${tenantId}] Metric pre-fetch complete — ${metricsStored} metrics, ${hostsWithData} hosts with data`);
  } else {
    console.error(`[org_analysis:${tenantId}] Metric pre-fetch FAILED — all hosts will have unknown efficiency labels:`, metricResult.reason);
  }

  if (metadataResult.status === "fulfilled") {
    const { hostsStored } = metadataResult.value;
    console.log(`[org_analysis:${tenantId}] Host metadata pre-fetch complete — ${hostsStored} hosts stored`);
  } else {
    console.error(`[org_analysis:${tenantId}] Host metadata pre-fetch FAILED — batch agents will fall back to per-host search_datadog_hosts:`, metadataResult.reason);
  }

  if (isAborted(runId)) {
    console.log(`[org_analysis:${tenantId}] Run aborted after metric pre-fetch — stopping`);
    return;
  }

  const batches: Array<typeof hosts> = [];
  for (let i = 0; i < hosts.length; i += BATCH_SIZE) {
    batches.push(hosts.slice(i, i + BATCH_SIZE));
  }

  console.log(`[org_analysis:${tenantId}] Found ${hosts.length} hosts → ${batches.length} batches, concurrency=${BATCH_CONCURRENCY}`);

  for (let i = 0; i < batches.length; i += BATCH_CONCURRENCY) {
    // Check abort signal before each wave
    if (isAborted(runId)) {
      console.log(`[org_analysis:${tenantId}] Run aborted — stopping at wave ${Math.floor(i / BATCH_CONCURRENCY) + 1}`);
      return;
    }

    const wave = batches.slice(i, i + BATCH_CONCURRENCY);
    console.log(`[org_analysis:${tenantId}] Wave ${Math.floor(i / BATCH_CONCURRENCY) + 1}/${Math.ceil(batches.length / BATCH_CONCURRENCY)} — running ${wave.length} batches`);
    await Promise.all(
      wave.map((batch, j) =>
        runHostBatchAgent(tenantId, batch, runId, i + j, batches.length)
      )
    );
  }

  // Final abort check before summarize
  if (isAborted(runId)) {
    console.log(`[org_analysis:${tenantId}] Run aborted — skipping summarize`);
    return;
  }

  console.log(`[org_analysis:${tenantId}] All batches complete, running summarize`);
  await runSummarizeAgent(tenantId, runId);
  console.log(`[org_analysis:${tenantId}] Org analysis complete`);
}
